scipy/interpolate/_rgi_cython.pyx

```diff
@@ -0,0 +1,164 @@
+            interval = -1
+    elif xval == b:
+        # Make the interval closed from the right
+        interval = nx - 2
+    else:
+        # Find the interval the coordinate is in
+        # (binary search with locality)
+        if xval >= x[interval]:
+            low = interval
+            high = nx - 2
+        else:
+            low = 0
+            high = interval
+
+        if xval < x[low+1]:
+            high = low
+
+        while low < high:
+            mid = (high + low)//2
+            if xval < x[mid]:
+                # mid < high
+                high = mid
+            elif xval >= x[mid + 1]:
+                low = mid + 1
+            else:
+                # x[mid] <= xval < x[mid+1]
+                low = mid
+                break
+
+        interval = low
+
+    return interval
+
+
+def evaluate_linear(values, indices, norm_distances, out_of_bounds):
+    # slice for broadcasting over trailing dimensions in self.values
+    vslice = (slice(None),) + (None,) * (values.ndim - len(indices))
+
+    # Compute shifting up front before zipping everything together
+    shift_norm_distances = [1 - yi for yi in norm_distances]
+    shift_indices = [i + 1 for i in indices]
+
+    # The formula for linear interpolation in 2d takes the form:
+    # values = self.values[(i0, i1)] * (1 - y0) * (1 - y1) + \
+    #          self.values[(i0, i1 + 1)] * (1 - y0) * y1 + \
+    #          self.values[(i0 + 1, i1)] * y0 * (1 - y1) + \
+    #          self.values[(i0 + 1, i1 + 1)] * y0 * y1
+    # We pair i with 1 - yi (zipped1) and i + 1 with yi (zipped2)
+    zipped1 = zip(indices, shift_norm_distances)
+    zipped2 = zip(shift_indices, norm_distances)
+
+    # Take all products of zipped1 and zipped2 and iterate over them
+    # to get the terms in the above formula. This corresponds to iterating
+    # over the vertices of a hypercube.
+    hypercube = itertools.product(*zip(zipped1, zipped2))
+    value = np.array([0.])
+    for h in hypercube:
+        edge_indices, weights = zip(*h)
+        weight = np.array([1.])
+        for w in weights:
+            weight = weight * w
+        value = value + np.asarray(values[edge_indices]) * weight[vslice]
+    return value
+
+
+@cython.wraparound(False)
+@cython.boundscheck(False)
+@cython.cdivision(True)
+def find_indices(grid, double[:, :] xi):
+    cdef long i, j, I, J, index, grid_i_size, nx
+    cdef double denom, grid_i_index
+    cdef np.ndarray[long, ndim=2] indices
+    cdef np.ndarray[double, ndim=2] norm_distances
+    cdef np.ndarray[double, ndim=1] grid_i
+
+    # find relevant edges between which xi are situated
+    indices = np.empty_like(xi, dtype=int)
+    # compute distance to lower edge in unity units
+    norm_distances = np.zeros_like(xi, dtype=float)
+
+    # indexes to iterate over
+    I = xi.shape[0]
+    J = xi.shape[1]
+
+    # iterate through dimensions
+    for i in range(I):
+        grid_i = np.asarray(grid[i], dtype=float)  # FIXME: do it earlier
```

Comment:
Indeed, can be done in the constructor!

Reply:
I'm not sure how to do this in the constructor. Can it be done via type hints?

Reply:
I *think* it's just a matter of calling asarray in `__init__`. Will take a look later today

Reply:
OK, it does call asarray in `__init__`, but it does not enforce `dtype=float`. So, something like this:

```diff
diff --git a/scipy/interpolate/_rgi.py b/scipy/interpolate/_rgi.py
index 6e1fc6d04e..06038e82e1 100644
--- a/scipy/interpolate/_rgi.py
+++ b/scipy/interpolate/_rgi.py
@@ -257,7 +257,7 @@ class RegularGridInterpolator:
             if not values.shape[i] == len(p):
                 raise ValueError("There are %d points and %d values in "
                                  "dimension %d" % (len(p), values.shape[i], i))
-        self.grid = tuple([np.asarray(p) for p in points])
+        self.grid = tuple([np.asarray(p, dtype=float) for p in points])
         self.values = values
 
     def __call__(self, xi, method=None):
@@ -310,6 +310,7 @@ class RegularGridInterpolator:
 
         xi_shape = xi.shape
         xi = xi.reshape(-1, xi_shape[-1])
+        xi = np.asarray(xi, dtype=float)
 
         # find nans in input
         nans = np.any(np.isnan(xi), axis=-1)
diff --git a/scipy/interpolate/_rgi_cython.pyx b/scipy/interpolate/_rgi_cython.pyx
index 1bb5a1e95c..a15e979b97 100644
--- a/scipy/interpolate/_rgi_cython.pyx
+++ b/scipy/interpolate/_rgi_cython.pyx
@@ -126,7 +126,7 @@ def evaluate_linear(values, indices, norm_distances, out_of_bounds):
 @cython.wraparound(False)
 @cython.boundscheck(False)
 @cython.cdivision(True)
-def find_indices(grid, double[:, :] xi):
+def find_indices(list grid not None, double[:, :] xi):
     cdef long i, j, I, J, index, grid_i_size, nx
     cdef double denom, grid_i_index
     cdef np.ndarray[long, ndim=2] indices
@@ -144,7 +144,7 @@ def find_indices(grid, double[:, :] xi):
 
     # iterate through dimensions
     for i in range(I):
-        grid_i = np.asarray(grid[i], dtype=float)  # FIXME: do it earlier
+        grid_i = grid[i]
         grid_i_size = grid_i.shape[0]
 
         for j in range(J):
```

Note that here I'm also typing `list grid not None`, but this is unlikely to be very relevant. Maybe cython will use `PyList_GET_ITEM` or some such and avoid a dance with indexing a general untyped object.